File: 2025-01-January-GFG-POTD/January 26 - Remove loop in Linked List/Solution.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

def removeLoop(head):
    if head is None or head.next is None:
        return

    slow, fast = head, head

    # Detecting the loop using Floyd's Cycle Detection Algorithm
    while fast and fast.next:
        slow = slow.next
        fast = fast.next.next

        # If slow and fast meet, there is a loop
        if slow == fast:
            break

    # If no loop is found, return
    if fast is None or fast.next is None:
        return

    # Find the start of the loop
    slow = head
    if slow != fast:
        while slow.next != fast.next:
            slow = slow.next
            fast = fast.next
    else:
        # Special case: when the loop starts at the head
        while fast.next != slow:
            fast = fast.next

    # Break the loop
    fast.next = None

# Helper function to create a new node
def newNode(data):
    return Node(data)

# Helper function to create a loop in the linked list
def makeLoop(head, x):
    if x == 0:
        return

    curr = head
    last = head

    currentPosition = 1
    while currentPosition < x:
        curr = curr.next
        currentPosition += 1

    while last.next:
        last = last.next
    last.next = curr

# Helper function to detect if a loop exists
def detectLoop(head):
    hare, tortoise = head.next, head
    while hare != tortoise:
        if hare is None or hare.next is None:
            return False
        hare = hare.next.next
        tortoise = tortoise.next
    return True

# Main function for testing
if __name__ == "__main__":
    n = int(input())
    arr = list(map(int, input().split()))
    pos = int(input())

    head = newNode(arr[0])
    tail = head
    for i in range(1, n):
        tail.next = newNode(arr[i])
        tail = tail.next

    makeLoop(head, pos)
    removeLoop(head)

    if detectLoop(head):
        print("false")
    else:
        print("true")

File: 2025-01-January-GFG-POTD/January 26 - Remove loop in Linked List/Solution.cpp
#include <iostream>
using namespace std;

struct Node {
    int data;
    Node* next;
};

// Function to remove the loop in the linked list
void removeLoop(Node* head) {
    if (head == nullptr || head->next == nullptr) return;

    Node *slow = head, *fast = head;

    // Detecting the loop using Floyd's Cycle Detection Algorithm
    while (fast != nullptr && fast->next != nullptr) {
        slow = slow->next;
        fast = fast->next->next;

        // If slow and fast meet, there is a loop
        if (slow == fast) break;
    }

    // If no loop is found, return
    if (fast == nullptr || fast->next == nullptr) return;

    // Find the start of the loop
    slow = head;
    if (slow != fast) {
        while (slow->next != fast->next) {
            slow = slow->next;
            fast = fast->next;
        }
    } else {
        // Special case: when the loop starts at the head
        while (fast->next != slow) {
            fast = fast->next;
        }
    }

    // Break the loop
    fast->next = nullptr;
}

// Helper function to create a new node
Node* newNode(int data) {
    Node* temp = new Node();
    temp->data = data;
    temp->next = nullptr;
    return temp;
}

// Helper function to create a loop in the linked list
void makeLoop(Node* head, int x) {
    if (x == 0) return;

    Node* curr = head;
    Node* last = head;

    int currentPosition = 1;
    while (currentPosition < x) {
        curr = curr->next;
        currentPosition++;
    }

    while (last->next != nullptr) last = last->next;
    last->next = curr;
}

// Helper function to detect if a loop exists
bool detectLoop(Node* head) {
    Node* hare = head->next;
    Node* tortoise = head;
    while (hare != tortoise) {
        if (hare == nullptr || hare->next == nullptr) return false;
        hare = hare->next->next;
        tortoise = tortoise->next;
    }
    return true;
}

// Main function for testing
int main() {
    int n, pos;
    cin >> n;

    int val;
    cin >> val;
    Node* head = newNode(val);
    Node* tail = head;

    for (int i = 1; i < n; ++i) {
        cin >> val;
        tail->next = newNode(val);
        tail = tail->next;
    }

    cin >> pos;
    makeLoop(head, pos);

    removeLoop(head);

    if (detectLoop(head)) {
        cout << "false\n";
    } else {
        cout << "true\n";
    }

    return 0;
}
